### PR TITLE
[binding] QPInverseProblemSolver: adds python binding

### PR DESCRIPTION
**Pull request by EulalieCoevoet**
Usage example:

```python
import Sofa.SoftRobotsInverse

class MyQPInverseProblemSolver(Sofa.SoftRobotsInverse.QPInverseProblemSolver):

    def __init__(self, *args, **kwargs):
        Sofa.SoftRobotsInverse.QPInverseProblemSolver.__init__(self, *args, **kwargs)

    def solveSystem(self):
        W = self.W()
        dfree = self.dfree()
        forces = self.lambda_force()  # pointer on lambda
        [...] # solve the problem and store the result in forces
        return True

def createScene(rootnode):
    rootnode.addObject(MyQPInverseProblemSolver)
    [...]
```

### CODE REVIEW

**Comment by damienmarchal** [comment]
I gave it a quick look and seems ok.

File: src/SoftRobots.Inverse/binding/Binding_QPInverseProblemSolver.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            auto& lambda = self.getConstraintProblem()->f;
+            return { lambda.ptr(), lambda.size()};
+        });
+
+    s.def("dfree", [](QPInverseProblemSolver& self) -> Eigen::Map<Eigen::Matrix<SReal, Eigen::Dynamic, 1> >
+        {
+            assert(self.getConstraintProblem());
+            auto& dfree = self.getConstraintProblem()->dFree;
+            return { dfree.ptr(), dfree.size()};
+        });
```
[the remark this answers is not in the document]
Comment (EulalieCoevoet) [reply src/SoftRobots.Inverse/binding/Binding_QPInverseProblemSolver.cpp:140]: I thought the same, but I didn't find why it does not work.

**Comment by alxbilger** [comment]
The unit test https://github.com/alxbilger/SofaPython3/blob/d16750f7fa6fe1075876877fc831f4d225ec9f90/bindings/Modules/tests/SofaConstraintSolver/matrix_access.py works for me with the following changes (without your PR):

```diff
                                                      "Sofa.Component.Mapping.MappedMatrix",
                                                      "Sofa.Component.Mass",
                                                      "Sofa.Component.ODESolver.Backward",
-                                                     "Sofa.Component.Topology.Container.Dynamic"])
+                                                     "Sofa.Component.Topology.Container.Dynamic",
+                                                     "SoftRobots.Inverse"])
 
         root.addObject("FreeMotionAnimationLoop", solveVelocityConstraintFirst=True)
-        root.addObject("GenericConstraintSolver", name="constraint_solver", tolerance=1e-9, maxIterations=1000)
+        root.addObject("QPInverseProblemSolver", name="constraint_solver", tolerance=1e-9, maxIterations=1000)
         root.addObject("StringMeshCreator", name="loader", resolution="20")
 
         root.addObject("EulerImplicitSolver")

```
Could you check that you have the same result?

**Comment by EulalieCoevoet** [comment]
> The unit test https://github.com/alxbilger/SofaPython3/blob/d16750f7fa6fe1075876877fc831f4d225ec9f90/bindings/Modules/tests/SofaConstraintSolver/matrix_access.py works for me with the following changes (without your PR).
> Could you check that you have the same result?

Yes it works. What does not work is when used like that (without duplicating what you did in the module SofaConstraintSolver):

```python
import Sofa.SoftRobotsInverse
import Sofa.SofaConstraintSolver


class MyQPInverseProblemSolver(Sofa.SoftRobotsInverse.QPInverseProblemSolver):

    def __init__(self, emio, *args, **kwargs):
        Sofa.SoftRobotsInverse.QPInverseProblemSolver.__init__(self, *args, **kwargs)

    def solveSystem(self):
        W = self.W()
        dfree = self.dfree()
        torques = self.lambda_force()
        return True
```

File: src/SoftRobots.Inverse/binding/Binding_QPInverseProblemSolver.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    m_currentCP->sendResults();
+}
+
+void moduleAddQPInverseProblemSolver(py::module &m)
+{
+    py::class_<softrobotsinverse::solver::QPInverseProblemSolver,
+               QPInverseProblemSolver_Trampoline,
+               BaseObject,
```
Comment (alxbilger) [inline src/SoftRobots.Inverse/binding/Binding_QPInverseProblemSolver.cpp:113]:
```suggestion
               ConstraintSolverImpl,
```
Try with this suggestion and remove the already-existing bindings

**Comment by alxbilger** [comment]
I am considering adding a trampoline class for `ConstraintSolverImpl`, not only for `QPInverseProblemSolver`. I see that your version needs specific data from `QPInverseProblemSolver` in `storeResults`. Would it be required to write its own `storeResults` for `ConstraintSolverImpl`?

File: src/SoftRobots.Inverse/binding/Binding_QPInverseProblemSolver.cpp (hunk 2 above)
Comment (EulalieCoevoet) [reply src/SoftRobots.Inverse/binding/Binding_QPInverseProblemSolver.cpp:113]: Yes that was the problem, thanks

**Review by alxbilger — APPROVED** [review]

**Pull request merged**